domain/model/watcher_test.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package model_test
+
+import (
+	"context"
+	"database/sql"
+	"testing"
+
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/core/changestream"
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/model"
+	"github.com/juju/juju/core/watcher/watchertest"
+	"github.com/juju/juju/domain"
+	"github.com/juju/juju/domain/model/service"
+	"github.com/juju/juju/domain/model/state"
+	modeltesting "github.com/juju/juju/domain/model/state/testing"
+	changestreamtesting "github.com/juju/juju/internal/changestream/testing"
+	loggertesting "github.com/juju/juju/internal/logger/testing"
+)
+
+type watcherSuite struct {
+	changestreamtesting.ControllerSuite
+}
+
+var _ = gc.Suite(&watcherSuite{})
+
+func Test(t *testing.T) {
+	gc.TestingT(t)
+}
+
+type dbModel struct {
+	UUID        string `db:"uuid"`
+	Activated   bool   `db:"activated"`
+	ModelTypeID int    `db:"model_type_id"`
+
+	Name      string `db:"name"`
+	CloudUUID string `db:"cloud_uuid"`
+	LifeID    int    `db:"life_id"`
+	OwnerUUID string `db:"owner_uuid"`
+}
+
+func (s *watcherSuite) TestWatchControllerDBModels(c *gc.C) {
+	logger := loggertesting.WrapCheckLog(c)
+	watchableDBFactory := changestream.NewWatchableDBFactoryForNamespace(s.GetWatchableDB, "model")
+	watcherFactory := domain.NewWatcherFactory(watchableDBFactory, logger)
+	st := state.NewState(func() (database.TxnRunner, error) { return watchableDBFactory() })
+	modelSvc := service.NewWatchableService(st, nil, loggertesting.WrapCheckLog(c), watcherFactory)
+
+	// Create a controller service watcher.
+	watcher, err := modelSvc.WatchActivatedModels(context.Background())
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(watcher, gc.NotNil)
+
+	modelName := "test-model"
+	var modelUUID model.UUID
+	var modelUUIDStr string
+
+	harness := watchertest.NewHarness(s, watchertest.NewWatcherC(c, watcher))
+
+	// Verifies that watchers do not receive any changes when newly unactivated models are created.
+	harness.AddTest(func(c *gc.C) {
+		// Create a new model named test-model and activate it.
+		modelUUID = modeltesting.CreateTestModel(c, s.TxnRunnerFactory(), modelName)
+		modelUUIDStr = modelUUID.String()
+
+		err := s.TxnRunner().StdTxn(context.Background(), func(ctx context.Context, tx *sql.Tx) error {
```

Comment:
You can't do this. Creating the test model has already fired the event in the transaction. I am surprised this works now.

Can I suggest starting over with this test and using the services to create the model's this time? It will be much more real word situation driven as these are integration tests to prove that under real world conditions we behave as expected.

Reply:
Sure, I have changed them to use the services instead